libgalois/src/DynamicBitset.cpp

```diff
@@ -96,90 +141,140 @@ katana::DynamicBitset::count() const {
   }
-
   // total num of set bits
-  Integer bitsetCount = tPrefixBitCounts[activeThreads - 1];
+  Integer bitset_count = thread_prefix_bit_counts[active_threads - 1];
 
   // calculate the indices of the set bits and save them to the offset
   // vector
-  if (bitsetCount > 0) {
-    size_t cur_size = offsets->size();
-    offsets->resize(cur_size + bitsetCount);
+  if (bitset_count > 0) {
+    size_t cur_size = set_elements->size();
+    set_elements->resize(cur_size + bitset_count);
+
     katana::on_each([&](unsigned tid, unsigned nthreads) {
-      auto [start, end] =
-          katana::block_range(size_t{0}, bitset.size(), tid, nthreads);
+      auto [start, end] = katana::block_range(
+          size_t{0}, underlying_bitvec.size(), tid, nthreads);
+
       Integer index = cur_size;
       if (tid != 0) {
-        index += tPrefixBitCounts[tid - 1];
+        index += thread_prefix_bit_counts[tid - 1];
       }
 
-      for (Integer i = start; i < end; ++i) {
-        if (bitset.test(i)) {
-          offsets->at(index) = i;
-          ++index;
+      for (uint64_t bitvec_index = start; bitvec_index < end; ++bitvec_index) {
+        // get set bits and add
+        uint64_t current_num = underlying_bitvec[bitvec_index];
+        uint64_t offset =
+            bitvec_index * katana::DynamicBitset::kNumBitsInUint64;
+
+        if (current_num == 0) {
+          // nothing to add
+          continue;
+        }
+
+        if (current_num == 1) {
+          // trailing 0s indicate location of set bit
+          set_elements->at(index++) = offset + CountTrailingZeroes(current_num);
+        } else if (current_num == std::numeric_limits<uint64_t>::max()) {
+          // add all
+          for (size_t i = 0; i < katana::DynamicBitset::kNumBitsInUint64; i++) {
+            set_elements->at(index++) = offset + i;
+          }
+        } else {
+          // add only set parts
+          for (size_t i = 0; i < katana::DynamicBitset::kNumBitsInUint64; i++) {
+            if (bitset.test(offset + i)) {
+              set_elements->at(index++) = offset + i;
+            }
+          }
         }
       }
     });
   }
 }
-}  //namespace
+
+template <typename Integer>
+void
+ComputeOffsetsSerial(
+    const katana::DynamicBitset& bitset, std::vector<Integer>* set_elements) {
+  const katana::PODVector<katana::DynamicBitset::TItem>& underlying_bitvec =
+      bitset.get_vec();
+  set_elements->reserve(bitset.SerialCount());
+
+  // loop through each int of the bitset invidiually
+  for (size_t bit_index = 0; bit_index < underlying_bitvec.size();
+       bit_index++) {
+    uint64_t int_to_examine = underlying_bitvec[bit_index];
+
+    if (int_to_examine == 0) {
+      // nothing set, skip
+      continue;
+    }
+
+    uint64_t offset = bit_index * katana::DynamicBitset::kNumBitsInUint64;
+
+    // TODO(l-hoang) optimize 63 set bits case? (find only 0, add the rest)
+    // optimize special corner cases case
+    if (int_to_examine == 1) {
```

Comment:
I'm confused by this - if this is integer equality don't you know that `int_to_examine == 0x0000000000000001` inside this block? Same comment for parallel version.

Reply:
yeah, this is a mistake on my part: int to examine used to be a count of the number of set bits but I changed it
incorrectly here